pkg/sql/gcjob_test/gcjob_test_test: TestSchemaChangeGCJob failed

pkg/sql/gcjob_test/gcjob_test_test.TestSchemaChangeGCJob [failed](https://tanzanite.cluster.engflow.com/invocations/default/4f9c3120-b2d9-47c1-b354-d3cc63fe6646?testReportRun=26&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvc3FsL2djam9iX3Rlc3Q6Z2Nqb2JfdGVzdF90ZXN0) on master @ [57bd0a92676b1c58ee7e29979ce3dceb5837abdb](https://github.com/cockroachdb/cockroach/commits/57bd0a92676b1c58ee7e29979ce3dceb5837abdb):

Fatal error:

```
panic: test timed out after 4m57s
running tests:
	TestSchemaChangeGCJob (4m57s)
	TestSchemaChangeGCJob/dropItem=0/ttlTime=1 (4m54s)

```
Stack: 

```
goroutine 339641 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2259 +0x3b9
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
* created by github.com/cockroachdb/cockroach/pkg/util/admission.(*WorkQueue).startClosingEpochs in goroutine 17790
* 	github.com/cockroachdb/cockroach/pkg/util/admission/work_queue.go:473 +0x4f
* 
* goroutine 155865 [select, 2 minutes]:
* github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFile).startTicker.func1()
* 	github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:171 +0xcb
* created by github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFile).startTicker in goroutine 18026
* 	github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:166 +0x58
* 
* goroutine 17924 [select, 2 minutes]:
* github.com/cockroachdb/cockroach/pkg/util/admission.initWorkQueue.func2()
* 	github.com/cockroachdb/cockroach/pkg/util/admission/work_queue.go:425 +0x85
* created by github.com/cockroachdb/cockroach/pkg/util/admission.initWorkQueue in goroutine 17790
* 	github.com/cockroachdb/cockroach/pkg/util/admission/work_queue.go:422 +0x3ce
* 
* goroutine 17927 [select]:
* github.com/cockroachdb/cockroach/pkg/util/admission.(*WorkQueue).startClosingEpochs.func1()
* 	github.com/cockroachdb/cockroach/pkg/util/admission/work_queue.go:502 +0x14d
* created by github.com/cockroachdb/cockroach/pkg/util/admission.(*WorkQueue).startClosingEpochs in goroutine 17790
* 	github.com/cockroachdb/cockroach/pkg/util/admission/work_queue.go:473 +0x4f
* 
* goroutine 17936 [sync.Cond.Wait, 2 minutes]:
* sync.runtime_notifyListWait(0xc003ade0a0, 0x0)
* 	GOROOT/src/runtime/sema.go:527 +0x159
* sync.(*Cond).Wait(0x0?)
* 	GOROOT/src/sync/cond.go:70 +0x85
* github.com/cockroachdb/pebble/record.(*flusherCond).Wait(...)
* 	github.com/cockroachdb/pebble/record/external/com_github_cockroachdb_pebble/record/log_writer.go:204
* github.com/cockroachdb/pebble/record.(*LogWriter).flushLoop(0xc003ade000, {0x4968ea0, 0x9cfe240})
* 	github.com/cockroachdb/pebble/record/external/com_github_cockroachdb_pebble/record/log_writer.go:438 +0x67c
* runtime/pprof.Do({0x658dfd8?, 0x9cfe240?}, {{0xc000139680?, 0xc0042a1aa0?, 0xc00204ef80?}}, 0xc00204efc0)
* 	GOROOT/src/runtime/pprof/runtime.go:51 +0x9d
* github.com/cockroachdb/pebble/record.NewLogWriter.func2()
* 	github.com/cockroachdb/pebble/record/external/com_github_cockroachdb_pebble/record/log_writer.go:357 +0x56
* created by github.com/cockroachdb/pebble/record.NewLogWriter in goroutine 17790
* 	github.com/cockroachdb/pebble/record/external/com_github_cockroachdb_pebble/record/log_writer.go:356 +0x416
* 
* goroutine 17930 [select, 2 minutes]:
* github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFile).startTicker.func1()
* 	github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:171 +0xcb
* created by github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFile).startTicker in goroutine 17790
* 	github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:166 +0x58
* 
* goroutine 26322 [select, 2 minutes]:
* database/sql.(*DB).connectionOpener(0xc00436ec30, {0x658e0b8, 0xc006a083c0})
* 	GOROOT/src/database/sql/sql.go:1218 +0x87
* created by database/sql.OpenDB in goroutine 17790
* 	GOROOT/src/database/sql/sql.go:791 +0x165
* 
*
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=26</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSchemaChangeGCJob.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-34997

--- a/pkg/sql/gcjob_test/gc_job_test.go
+++ b/pkg/sql/gcjob_test/gc_job_test.go
@@ -86,7 +86,6 @@ func TestSchemaChangeGCJob(t *testing.T) {
 
 func doTestSchemaChangeGCJob(t *testing.T, dropItem DropItem, ttlTime TTLTime) {
 	blockGC := make(chan struct{}, 1)
-	defer close(blockGC)
 	params := base.TestServerArgs{}
 	params.ScanMaxIdleTime = time.Millisecond
 	params.Knobs.JobsTestingKnobs = jobs.NewTestingKnobsWithShortIntervals()
@@ -99,6 +98,10 @@ func doTestSchemaChangeGCJob(t *testing.T, dropItem DropItem, ttlTime TTLTime) {
 	s, db, kvDB := serverutils.StartServer(t, params)
 	ctx := context.Background()
 	defer s.Stopper().Stop(ctx)
+	// The deferred call to unblock the GC job needs to run before the deferred
+	// call to stop the TestServer. Otherwise, the quiesce step of shutting down
+	// can hang forever waiting for the GC job.
+	defer close(blockGC)
 	sqlDB := sqlutils.MakeSQLRunner(db)
 
 	sqlDB.Exec(t, `SET CLUSTER SETTING sql.gc_job.wait_for_gc.interval = '1s';`)

